### PR TITLE
Add `log_audio`, `TBAudio`, tests, examples

### PR DESCRIPTION
**Pull request by shashikdm**

### CODE REVIEW

File: src/Loggers/LogText.jl [hunk 1]
```diff
@@ -25,6 +25,9 @@ function text_summary(name::String, text::Any)
     #content of the tensor
     textstringval = Vector{Array{UInt8, 1}}()
     if isa(text, AbstractArray)
+        if ndims(text) == 2
+            text = permutedims(text, (2, 1))
+        end
```
Comment (PhilipVinc) [inline src/Loggers/LogText.jl:30]: What is this for?

Comment (shashikdm) [reply src/Loggers/LogText.jl:30]: Julia is column major. but tensorboard assumes row major and that creates problems when actually logging matrices as text. after transposing it works as expected.

File: src/logger_dispatch_overrides.jl [hunk 2]
```diff
@@ -24,6 +24,31 @@ preprocess(name, val::WrapperLogType, data) = push!(data, name=>val)
 Base.show(io::IO, mime::AbstractString, x::WrapperLogType) =
     Base.show(io, mime, content(x))
 
+########## For things going to LogAudio ##############################
+"""
+    TBAudio(data, samplerate)
```
Comment (PhilipVinc) [inline src/logger_dispatch_overrides.jl]: wrong docstring? It's supposed to be for `TBAudios`

Comment (shashikdm) [reply src/logger_dispatch_overrides.jl]: I will change that

File: src/logger_dispatch_overrides.jl [hunk 3]
```diff
@@ -24,6 +24,31 @@ preprocess(name, val::WrapperLogType, data) = push!(data, name=>val)
...
+function preprocess(name, val::TBAudios, data)
+    for (n, audio) in enumerate(val.data)
+        preprocess(name*"/$n", TBAudio(audio, val.samplerate), data)
+    end
+end
+
+struct TBAudio <: WrapperLogType
+    data::AbstractArray
+    samplerate::Real
+end
```
Comment (PhilipVinc) [inline src/logger_dispatch_overrides.jl:52]: Can you add a docstring here?

**Comment by PhilipVinc** [comment]
Do we really need to ship a 1 Mb `.wav` file only for testing? 
Can't we just generate some random noisy data and be done with it?

File: src/logger_dispatch_overrides.jl (hunk 3 above)
Comment (shashikdm) [reply src/logger_dispatch_overrides.jl:52]: yes I will add now

**Comment by shashikdm** [comment]
No problem with noisy data. I will remove the test file now

File: src/Loggers/LogAudio.jl [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+"""
+    log_audio(logger::TBLogger, name::AbstractString, samples::AbstractArray, samplerate::Real, step)
+
+Logs an audio clip with name `name` at step `step`
+- samples: Array of samples N*C where N = number of samples and C = number of channels
+- samplerate: Sampling rate or Sampling frequency: a Real value
+"""
+function log_audio(logger::TBLogger, name::AbstractString, samples::AbstractArray, samplerate::Real; step=nothing)
+    summ    = SummaryCollection()
+    SummaryCollection(audio_summary(name, samples, samplerate))
```
Comment (oxinabox) [inline src/Loggers/LogAudio.jl]: You forgot to assign this to a variable

Comment (shashikdm) [reply src/Loggers/LogAudio.jl]: 🤦😅. fixed now.

**Comment by oxinabox** [comment]
@shashikdm a bit off-topic, but you might want to learn to use `git rebase`.
Rather than repeatedly `merge`ing in *master*, `rebase`ing on top of *master*,
gives a much cleaner git history and makes reviewing PRs much easier.

**Comment by shashikdm** [comment]
@oxinabox I will take care from next time. Is there any DataType in Julia for sound type variables (like `Colorant` for Images) that you know of? I looked into https://github.com/JuliaAudio/ but could not find any. Currently `log_audio` supports AbstractArray of samples. if there's a particular datatype then it will help in automatic dispatch.

**Comment by oxinabox** [comment]
You should be able to do the same FileIO thing that was done for Images,
and that will let us also do the same thing to avoid direct dependencies
as in
https://github.com/PhilipVinc/TensorBoardLogger.jl/pull/28

**Comment by oxinabox** [comment]
> @oxinabox I will take care from next time. Is there any DataType in Julia for sound type variables (like Colorant for Images) that you know of? I looked into https://github.com/JuliaAudio/ but could not find any. Currently log_audio supports AbstractArray of samples. if there's a particular datatype then it will help in automatic dispatch.

I'm not sure, perhaps @ssfrr will be able to answer?

**Comment by shashikdm** [comment]
yes I had dropped him a dm in slack a few days ago. He has not been online I guess.

**Comment by ssfrr** [comment]
Yeah, I've been on vacation and not on slack. The JuliaAudio stack stores audio as the `SampleBuf` type (from [SampledSignals.jl](https://github.com/JuliaAudio/SampledSignals.jl), which is basically a wrapper around an `AbstractArray` that also stores a samplerate.

Let me know if there's any more info you need.

**Comment by shashikdm** [comment]
Great! thank you very much

**Comment by shashikdm** [comment]
So we'll have to add a dependency to `SampledSignals.jl`. Then we can implement Logger despatch for `SampleBuf` type variables.

**Comment by oxinabox** [comment]
SampledSignals does not have a *huge* set of dependencies.  though it has more than is ideal.
Particularly since I imagine the use of TensorBoard for audio will be rare.

How about we open an issue and think about that later.
See if anyone ends up using TensorBoard logger with this this type.
(and if so, compatibility can come from either package, or from the user's script (though that is technically piracy))

**Comment by PhilipVinc** [comment]
Otherwise, we could create a submodule so that to log audio files you must first do `using TensorBoardLogger.Audio`. We will still need to depend on `SampledSignals`, but we will only load it and it's dependency if needed.

Though this somewhat defeats the purpose of a general logging framework, because you need to declare in advance that you will be logging audio files...

**Comment by oxinabox** [comment]
Lets save working that out for #29 and when it comes up in practice.
Lets get this PR going for now

**Comment by PhilipVinc** [comment]
Agreed. 
@shashikdm  If we leave out the dependency on `SampledSignals`, what's the status?

**Comment by shashikdm** [comment]
> @shashikdm If we leave out the dependency on `SampledSignals`, what's the status?

`log_audios` and  `load_audio` is usable for explicit interface.  
`TBAudios` and `TBAudio` can be used as wrappers for logger interface.  
However, there is no automatic dispatch to `audio_summary`.

**Comment by PhilipVinc** [comment]
That's okay for now. 
Can I merge?

**Comment by shashikdm** [comment]
Yes, you can merge

**Pull request merged**